### PR TITLE
Non-zero exit code for failing Karma; Fixed browser execute boolean f…

### PR DESCRIPTION
**Pull request by mattalexx**
…lags from CI

These issues were brought up here:

> @Gusted This needs tweaking, it looks like.
> 
> * This is the ESR run, and Chrome should not be used. [darkreader/darkreader/runs/5591884298?check_suite_focus=true#step:10:259](https://github.com/darkreader/darkreader/runs/5591884298?check_suite_focus=true#step:10:259)
> * Firefox is failing but the test run is succeeding: [darkreader/darkreader/runs/5591884298?check_suite_focus=true#step:10:472](https://github.com/darkreader/darkreader/runs/5591884298?check_suite_focus=true#step:10:472)
> 
> I'll figure it out and post a PR.

https://github.com/darkreader/darkreader/pull/8393#issuecomment-1072097576

As a change of pace, I am removing more lines than I'm adding in this PR :-)

### CODE REVIEW

File: .github/workflows/test-browser.yml [hunk 1]
```diff
@@ -57,7 +57,7 @@ jobs:
         if: ${{ matrix.chrome }}
         run: |
           chrome --version
-          echo CHROME_BIN="$(which chrome)" | tee -a "$GITHUB_ENV"
+          { echo CHROME_BIN="$(which chrome)"; echo CHROME_TEST=1; } | tee -a "$GITHUB_ENV"
```
Comment (mattalexx) [inline .github/workflows/test-browser.yml:60]: In the final iteration, I was using Karma's `CHROME_BIN` var as the "test this browser" flag. But I wasn't aware that that var is automatically set to the pre-installed Chrome version on the GitHub Actions machine. So I added a custom var, `CHROME_TEST` (and `FIREFOX_TEST`). Fixed.

File: package.json [hunk 2]
```diff
@@ -21,8 +21,8 @@
     "test:ci": "npm run test:unit",
     "test:coverage": "jest --config=tests/unit/jest.config.js --coverage",
     "test:firefox": " npm run debug && jest --config=tests/browser/jest.config.firefox.js --runInBand",
-    "test:inject": "node --max-old-space-size=3072 tests/inject/run.js",
-    "test:inject:debug": "node --max-old-space-size=3072 tests/inject/run.js --debug",
+    "test:inject": "node --max-old-space-size=3072 node_modules/.bin/karma start ./tests/inject/karma.conf.js",
+    "test:inject:debug": "node --max-old-space-size=3072 node_modules/.bin/karma start ./tests/inject/karma.conf.js --debug",
```
Comment (mattalexx) [inline package.json:25]: Onto the problem with failing tests returning a zero and leaving the CI test run green.

The signals and cause and effect was getting confusing in `./tests/inject/run.js`, with `process`, `karmaServer`, and `echoServer` all being eventEmitters and all affecting one another. 

The cleanest way I found to handle it was to implement the echo server as a Karma middleware plugin, delete `run.js`, and use `karma start` in the package.json script.

File: tests/inject/karma.conf.js [hunk 3]
```diff
@@ -6,22 +9,40 @@ const rollupPluginReplace = require('@rollup/plugin-replace');
...
 /**
- * @param   {import('karma').Config} config
- * @returns {import('karma').ConfigOptions}
+ * @param {LocalConfig} config
+ * @param {Record<string, string>} env
+ * @returns {ConfigOptions}
  */
-function configureKarma(config) {
-    const headless = config.headless || process.env.KARMA_HEADLESS || false;
+function configureKarma(config, env) {
```
Comment (mattalexx) [inline tests/inject/karma.conf.js:19]: I set up a test for this method, so I parameterized the `process.env` global here.

File: tests/inject/karma.conf.js [hunk 4]
```diff
@@ -6,22 +9,40 @@ const rollupPluginReplace = require('@rollup/plugin-replace');
...
+    /** @type {ConfigOptions} */
     let options = {
+        failOnFailingTestSuite: true,
+        failOnEmptyTestSuite: true,
         basePath: '../..',
         frameworks: ['jasmine'],
         files: [
             'tests/inject/support/customize.ts',
             'tests/inject/support/polyfills.ts',
+            'tests/inject/dynamic/link-override.tests.ts',
```
Comment (mattalexx) [inline tests/inject/karma.conf.js:31]: Whoops, dev cruft.

File: tests/inject/karma.conf.js [hunk 5]
```diff
@@ -6,22 +9,40 @@ const rollupPluginReplace = require('@rollup/plugin-replace');
...
             {pattern: 'tests/inject/**/*.tests.ts', watched: false},
         ],
+        plugins: [
+            'karma-chrome-launcher',
+            'karma-coverage',
+            'karma-firefox-launcher',
+            'karma-rollup-preprocessor',
+            'karma-jasmine',
+            'karma-spec-reporter',
+            karmaPluginEchoServer,
```
Comment (mattalexx) [inline tests/inject/karma.conf.js:41]: As you're problably aware, Karma has been implicitly loading all Karma plugins by just loading `./node_modules/karma-*/`. 

Unfortunately this is not possible for the echo server plugin unless I set it up as an actual package, which I decided was overkill. But if I add `karmaPluginEchoServer` to the list, Karma complains until I add everything to the list.

File: tests/inject/karma.conf.js [hunk 6]
```diff
@@ -102,9 +130,13 @@ function configureKarma(config) {
...
  * @returns {void}
  */
 module.exports = (config) => {
-    config.set(configureKarma(config));
+    config.set(configureKarma(config, process.env));
 };
+
+if (process.env.NODE_ENV === 'test') {
+    module.exports.configureKarma = configureKarma;
+}
```
Comment (mattalexx) [inline tests/inject/karma.conf.js:142]: Test-only export.

File: tests/inject/karma.conf.js (hunk 4 above)
Comment (mattalexx) [reply tests/inject/karma.conf.js]: Fixed.

**Review by Gusted — APPROVED** [review]

**Pull request merged**